### ISSUE TITLE
kafka module invalid CA certificate

### ISSUE THREAD

**Issue by dborysenko** [fluent/fluent-plugin-kafka#305]
For those who struggling make Fluentd work with kafka cluster over SSL using self signed rootCA as I did:

Regardless of what "ssl_verify_hostname" is set to, I was getting below errors:

`2019-12-10 23:23:06 +0000 [warn]: #0 failed to flush the buffer. retry_time=2 next_retry_seconds=2019-12-10 23:23:08 +0000 chunk="59961c75a8edc4c9353eb69c3fe14844" error_class=OpenSSL::SSL::SSLError error="SSL_connect returned=1 errno=0 state=error: certificate verify failed (invalid CA certificate)"`

It appears starting version 1.1.1, openssl will fail certificate verification step if below two rules are not met:
1. self-signed CA certificate should be generated having below extensions:
        X509v3 extensions:
            X509v3 Key Usage:
                Digital Signature, Certificate Sign, CRL Sign
            X509v3 Basic Constraints:
                CA:TRUE

2. server certificate signed with rootCA should have below extensions set:
        X509v3 extensions:
            _X509v3 Subject Alternative Name:
                IP Address:192.168.02, DNS:mykafka.host.example.com_
            X509v3 Key Usage:
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage:
                TLS Web Server Authentication
            X509v3 Basic Constraints:
                CA:FALSE

SAN section is needed for "ssl_verify_hostname=true" only.

Thought this might help someone, if it is not an appropriate place to put, please move or drop it.

**Comment by milonjames** [comment]
@dborysenko We are seeing a different error while trying to make FluentD Kafka output use SSL and SASL auth. Wondering whether you faced the same issue and it would be helpful for us to get some clue around this error.

==

Send exception occurred: SSL_connect returned=1 errno=0 state=error: wrong version number

====

**Fluent version 1.8
fluent-plugin-kafka version v0.12.3**

The error looks like it's saying Fluentd is trying a TLS version which is not supported with Kafka, however not sure wether that is really the case.

**Comment by dborysenko** [comment]
Hi @milonjames,
As you saw we had a little bit different error. With your stuff I’d start with capturing traffic between fluent and kafka nodes. It will show you if the error is on SSL transport layer.

**Comment by mikolajmorawski** [comment]
Hi @milonjames 
I encounter the same issue yesterday. It is related to the Kafka keystore file. The keystore should be created with the "-keyalg RSA" parameter. I had the DSA in the Kafka keystore and it was my root cause for the "wrong version number" error.

I found the solution here:
https://github.com/openssl/openssl/issues/6289 
 
tl; dr - when creating the keystore, make sure to use "-keyalg RSA"

Additionally, I am facing strange behavior after fluentd kafka plugin upgrade to the latest version. 

Please check the issue that I created today:
https://github.com/fluent/fluent-plugin-kafka/issues/325

The old version of this plugin is working fine.

**Comment by milonjames** [comment]
@mikolajmorawski Thanks for this finding and sharing it. I confirm this is the case after reading https://wiki.openssl.org/index.php/TLS1.3. Let me try reconfiguring the Kafka keystore to use RSA algorithm.

**Comment by tsrivishnu** [comment]
I was facing the same issue with a different Kafka client and I confirm that using `-keyalg RSA` solved it.

**Comment by milonjames** [comment]
@tsrivishnu Yes, for me also that worked.
